### ISSUE TITLE
Tests fail/hang with Python 3.12

### ISSUE THREAD

**Issue by mgorny** [pypa/distlib#199]
**Describe the bug**
When running the test suite with Python 3.12.0b1, the test suite reaches the following test and then hangs:

```
test_download (test_index.PackageIndexTestCase.test_download) ... ERROR
```

**To Reproduce**
1. `tox -e py312`

**Expected behavior**
Tests passing ;-).

**Screenshots**
n/a

**Environment**
 - OS, including version: Gentoo Linux
 - Version of this library: 3.6.0, git 05375908c1b2d6b0e74bdeb574569d3609db9f56

**Additional information**
n/a

**Comment by mgorny** [comment]
Not sure how helpful is that but after running the test suite via pytest with pytest-timeout plugin, I've managed to get the following error:

```pytb
============================================================== FAILURES ===============================================================
_________________________________________________ PackageIndexTestCase.test_download __________________________________________________

self = <test_index.PackageIndexTestCase testMethod=test_download>

    @unittest.skipIf(IN_GITHUB_WORKFLOW, 'This test is end-of-line dependent')
    def test_download(self):  # pragma: no cover
        digest = '913093474942c5a564c011f232868517' # for testsrc/README.txt
        certfile = os.path.join(HERE, 'keycert.pem')
        server = self.make_https_server(certfile)
        url = 'https://localhost:%d/README.txt' % server.port
        fd, fn = tempfile.mkstemp()
        os.close(fd)
        self.addCleanup(os.remove, fn)
        with open(os.path.join(HERE, 'testsrc', 'README.txt'), 'rb') as f:
            data = f.read()
        self.index.ssl_verifier = HTTPSHandler(certfile)
>       self.index.download_file(url, fn)   # no digest

tests/test_index.py:333: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
distlib/index.py:403: in download_file
    sfp = self.send_request(Request(url))
distlib/index.py:456: in send_request
    return opener.open(req)
/usr/lib/python3.12/urllib/request.py:515: in open
    response = self._open(req, data)
/usr/lib/python3.12/urllib/request.py:532: in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
/usr/lib/python3.12/urllib/request.py:492: in _call_chain
    result = func(*args)
distlib/util.py:1481: in https_open
    return self.do_open(self._conn_maker, req)
/usr/lib/python3.12/urllib/request.py:1344: in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
/usr/lib/python3.12/http/client.py:1320: in request
    self._send_request(method, url, body, headers, encode_chunked)
/usr/lib/python3.12/http/client.py:1366: in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
/usr/lib/python3.12/http/client.py:1315: in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
/usr/lib/python3.12/http/client.py:1075: in _send_output
    self.send(msg)
/usr/lib/python3.12/http/client.py:1019: in send
    self.connect()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <distlib.util.HTTPSConnection object at 0x7f97eb0240b0>

    def connect(self):
        sock = socket.create_connection((self.host, self.port), self.timeout)
        if getattr(self, '_tunnel_host', False):
            self.sock = sock
            self._tunnel()
    
        context = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
        if hasattr(ssl, 'OP_NO_SSLv2'):
            context.options |= ssl.OP_NO_SSLv2
>       if self.cert_file:
E       AttributeError: 'HTTPSConnection' object has no attribute 'cert_file'

distlib/util.py:1438: AttributeError
--------------------------------------------------------- Captured log setup ----------------------------------------------------------
WARNING  test_index:test_index.py:107 PyPI test server could not be run
-------------------------------------------------------- Captured stderr call ---------------------------------------------------------

+++++++++++++++++++++++++++++++++++ Timeout ++++++++++++++++++++++++++++++++++++

~~~~~~~~~~~~~~~~~~~~~ Stack of Thread-1 (140290459121344) ~~~~~~~~~~~~~~~~~~~~~~
  File "/usr/lib/python3.12/threading.py", line 1009, in _bootstrap
    self._bootstrap_inner()
  File "/usr/lib/python3.12/threading.py", line 1052, in _bootstrap_inner
    self.run()
  File "/tmp/distlib/tests/support.py", line 253, in run
    self.server.serve_forever(0.05)
  File "/usr/lib/python3.12/socketserver.py", line 240, in serve_forever
    self._handle_request_noblock()
  File "/usr/lib/python3.12/socketserver.py", line 313, in _handle_request_noblock
    request, client_address = self.get_request()
  File "/tmp/distlib/tests/support.py", line 223, in get_request
    sock = context.wrap_socket(sock, server_side=True)
  File "/usr/lib/python3.12/ssl.py", line 455, in wrap_socket
    return self.sslsocket_class._create(
  File "/usr/lib/python3.12/ssl.py", line 1013, in _create
    self.do_handshake()
  File "/usr/lib/python3.12/ssl.py", line 1284, in do_handshake
    self._sslobj.do_handshake()

+++++++++++++++++++++++++++++++++++ Timeout ++++++++++++++++++++++++++++++++++++
_____________________________________________ PackageIndexTestCase.test_ssl_verification ______________________________________________

self = <test_index.PackageIndexTestCase testMethod=test_ssl_verification>

    def test_ssl_verification(self):
        certfile = os.path.join(HERE, 'keycert.pem')
        server = self.make_https_server(certfile)
        url = 'https://localhost:%d/' % server.port
        req = Request(url)
        self.index.ssl_verifier = HTTPSHandler(certfile)
>       response = self.index.send_request(req)

tests/test_index.py:318: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
distlib/index.py:456: in send_request
    return opener.open(req)
/usr/lib/python3.12/urllib/request.py:515: in open
    response = self._open(req, data)
/usr/lib/python3.12/urllib/request.py:532: in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
/usr/lib/python3.12/urllib/request.py:492: in _call_chain
    result = func(*args)
distlib/util.py:1481: in https_open
    return self.do_open(self._conn_maker, req)
/usr/lib/python3.12/urllib/request.py:1344: in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
/usr/lib/python3.12/http/client.py:1320: in request
    self._send_request(method, url, body, headers, encode_chunked)
/usr/lib/python3.12/http/client.py:1366: in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
/usr/lib/python3.12/http/client.py:1315: in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
/usr/lib/python3.12/http/client.py:1075: in _send_output
    self.send(msg)
/usr/lib/python3.12/http/client.py:1019: in send
    self.connect()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <distlib.util.HTTPSConnection object at 0x7f97eb0246b0>

    def connect(self):
        sock = socket.create_connection((self.host, self.port), self.timeout)
        if getattr(self, '_tunnel_host', False):
            self.sock = sock
            self._tunnel()
    
        context = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
        if hasattr(ssl, 'OP_NO_SSLv2'):
            context.options |= ssl.OP_NO_SSLv2
>       if self.cert_file:
E       AttributeError: 'HTTPSConnection' object has no attribute 'cert_file'

distlib/util.py:1438: AttributeError
-------------------------------------------------------- Captured stderr call ---------------------------------------------------------

+++++++++++++++++++++++++++++++++++ Timeout ++++++++++++++++++++++++++++++++++++

~~~~~~~~~~~~~~~~~~~~~ Stack of Thread-2 (140290441291456) ~~~~~~~~~~~~~~~~~~~~~~
  File "/usr/lib/python3.12/threading.py", line 1009, in _bootstrap
    self._bootstrap_inner()
  File "/usr/lib/python3.12/threading.py", line 1052, in _bootstrap_inner
    self.run()
  File "/tmp/distlib/tests/support.py", line 253, in run
    self.server.serve_forever(0.05)
  File "/usr/lib/python3.12/socketserver.py", line 240, in serve_forever
    self._handle_request_noblock()
  File "/usr/lib/python3.12/socketserver.py", line 313, in _handle_request_noblock
    request, client_address = self.get_request()
  File "/tmp/distlib/tests/support.py", line 223, in get_request
    sock = context.wrap_socket(sock, server_side=True)
  File "/usr/lib/python3.12/ssl.py", line 455, in wrap_socket
    return self.sslsocket_class._create(
  File "/usr/lib/python3.12/ssl.py", line 1013, in _create
    self.do_handshake()
  File "/usr/lib/python3.12/ssl.py", line 1284, in do_handshake
    self._sslobj.do_handshake()

~~~~~~~~~~~~~~~~~~~~~ Stack of Thread-1 (140290459121344) ~~~~~~~~~~~~~~~~~~~~~~
  File "/usr/lib/python3.12/threading.py", line 1009, in _bootstrap
    self._bootstrap_inner()
  File "/usr/lib/python3.12/threading.py", line 1052, in _bootstrap_inner
    self.run()
  File "/tmp/distlib/tests/support.py", line 253, in run
    self.server.serve_forever(0.05)
  File "/usr/lib/python3.12/socketserver.py", line 240, in serve_forever
    self._handle_request_noblock()
  File "/usr/lib/python3.12/socketserver.py", line 313, in _handle_request_noblock
    request, client_address = self.get_request()
  File "/tmp/distlib/tests/support.py", line 223, in get_request
    sock = context.wrap_socket(sock, server_side=True)
  File "/usr/lib/python3.12/ssl.py", line 455, in wrap_socket
    return self.sslsocket_class._create(
  File "/usr/lib/python3.12/ssl.py", line 1013, in _create
    self.do_handshake()
  File "/usr/lib/python3.12/ssl.py", line 1284, in do_handshake
    self._sslobj.do_handshake()

+++++++++++++++++++++++++++++++++++ Timeout ++++++++++++++++++++++++++++++++++++
```

**Comment by mgorny** [comment]
Ah, thanks, my bad for not searching through pull requests.

**Comment by vsajip** [comment]
I've merged it now, so you can try with the latest version from this repository in your environment - please report your results.

**Comment by mgorny** [comment]
All green now, thanks!

**Comment by vsajip** [comment]
Closed via #198.
